### PR TITLE
dev: add Providers for each server

### PR DESCRIPTION
**Pull request by eugypalu**
Resolves: #1277 

# Pull Request type
Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

This PR introduces abstractions for the Alchemy, Debug, and Pool RPC providers, similar to the existing EthProvider abstraction

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by tcoratger** [comment]
@greged93 I like the idea here, on the PR mempool I just encountered provider problems so I like this split. I think however that we should wait to merge the mempool before merging that right?

File: src/providers/alchemy_provider/provider.rs [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+impl<P: EthereumProvider> AlchemyStruct<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for AlchemyStruct<P> {
```
Comment (greged93) [inline src/providers/alchemy_provider/provider.rs:52]: Why do we need `EthereumProvider` for the `AlchemyStruct`?

File: src/providers/alchemy_provider/provider.rs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    async fn token_allowance(
+        &self,
+        contract_address: Address,
+        owner: Address,
+        spender: Address,
+    ) -> EthProviderResult<U256>;
+}
+
+#[derive(Debug, Clone)]
+pub struct AlchemyStruct<P: EthereumProvider> {
```
Comment (greged93) [inline src/providers/alchemy_provider/provider.rs:41]: I think we need to rename, maybe into `AlchemyDataProvider` to follow the same naming as the eth provider

File: src/providers/debug_provider/provider.rs [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+impl<P: EthereumProvider> DebugStruct<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for DebugStruct<P> {
```
Comment (greged93) [inline src/providers/debug_provider/provider.rs:68]: same, why do we need this?

File: src/providers/debug_provider/provider.rs [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    async fn trace_call(
+        &self,
+        request: TransactionRequest,
+        block_number: Option<BlockId>,
+        opts: Option<GethDebugTracingCallOptions>,
+    ) -> EthProviderResult<GethTrace>;
+}
+
+#[derive(Debug, Clone)]
+pub struct DebugStruct<P: EthereumProvider> {
```
Comment (greged93) [inline src/providers/debug_provider/provider.rs:57]: also rename

File: src/providers/alchemy_provider/contracts/erc20.rs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 #![allow(clippy::pub_underscore_fields)]
```
Comment (greged93) [inline src/providers/alchemy_provider/contracts/erc20.rs:1]: this adds a lot of boiler plate code, let's make a single contract folder

File: src/providers/pool_provider/provider.rs [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+impl<P: EthereumProvider> PoolStruct<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for PoolStruct<P> {
```
Comment (greged93) [inline src/providers/pool_provider/provider.rs:34]: same, do we need this?

File: src/providers/pool_provider/provider.rs [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+#[auto_impl(Arc, &)]
+pub trait PoolProvider {
+    async fn txpool_status(&self) -> EthProviderResult<TxpoolStatus>;
+    async fn txpool_inspect(&self) -> EthProviderResult<TxpoolInspect>;
+    async fn txpool_content_from(&self, from: Address) -> EthProviderResult<TxpoolContentFrom>;
+    async fn txpool_content(&self) -> EthProviderResult<TxpoolContent>;
+}
+
+#[derive(Debug, Clone)]
+pub struct PoolStruct<P: EthereumProvider> {
```
Comment (greged93) [inline src/providers/pool_provider/provider.rs:23]: same, rename

File: src/eth_rpc/servers/alchemy_rpc.rs [hunk 8]
```diff
@@ -1,78 +1,42 @@
...
+    providers::alchemy_provider::provider::AlchemyProvider,
 };
-use futures::future::join_all;
-use jsonrpsee::core::{async_trait, RpcResult};
-use reth_primitives::{Address, BlockId, BlockNumberOrTag, U256};
+use async_trait::async_trait;
+use jsonrpsee::core::RpcResult as Result;
+use reth_primitives::{Address, U256};
 
-/// The RPC module for the Ethereum protocol required by Kakarot.
```
Comment (greged93) [inline src/eth_rpc/servers/alchemy_rpc.rs:12]: is there a reason for removing the comments?

File: src/eth_rpc/servers/alchemy_rpc.rs [hunk 9]
```diff
@@ -1,78 +1,42 @@
...
 #[derive(Debug)]
-pub struct AlchemyRpc<P: EthereumProvider> {
-    /// The provider for interacting with the Ethereum network.
-    eth_provider: P,
+pub struct AlchemyRpc<P: AlchemyProvider> {
+    provider: P,
```
Comment (greged93) [inline src/eth_rpc/servers/alchemy_rpc.rs:14]: can we do `alchemy_provider` and use `AP` as the generic?

File: src/eth_rpc/servers/alchemy_rpc.rs [hunk 10]
```diff
@@ -1,78 +1,42 @@
...
-            }))
-            .await
-            .into_iter()
-            .collect::<Result<Vec<_>, EthApiError>>()?,
-        })
+impl<P> AlchemyApiServer for AlchemyRpc<P>
+where
+    P: AlchemyProvider + Send + Sync + 'static,
+{
+    async fn token_balances(&self, address: Address, contract_addresses: Vec<Address>) -> Result<TokenBalances> {
```
Comment (greged93) [inline src/eth_rpc/servers/alchemy_rpc.rs:31]: please don't remove the `tracing` macro

File: src/eth_rpc/servers/debug_rpc.rs [hunk 11]
```diff
@@ -1,160 +1,62 @@
...
     trace::geth::{GethDebugTracingCallOptions, GethDebugTracingOptions, GethTrace, TraceResult},
     BlockId, BlockNumberOrTag, TransactionRequest,
 };
-use std::sync::Arc;
 
 /// The RPC module for the implementing Net api
 #[derive(Debug)]
-pub struct DebugRpc<P: EthereumProvider> {
-    eth_provider: P,
+pub struct DebugRpc<P: DebugProvider> {
```
Comment (greged93) [inline src/eth_rpc/servers/debug_rpc.rs:12]: `debug_provider` and `DP`?

File: src/eth_rpc/servers/txpool_rpc.rs [hunk 12]
```diff
@@ -1,32 +1,37 @@
...
-use reth_rpc_types::txpool::{TxpoolContent, TxpoolContentFrom, TxpoolInspect, TxpoolInspectSummary, TxpoolStatus};
+use reth_rpc_types::txpool::{TxpoolContent, TxpoolContentFrom, TxpoolInspect, TxpoolStatus};
 use tracing::instrument;
 
 /// The RPC module for implementing the Txpool api
 #[derive(Debug)]
-pub struct TxpoolRpc<P: EthereumProvider> {
-    eth_provider: P,
+pub struct TxpoolRpc<P: PoolProvider> {
+    provider: P,
```
Comment (greged93) [inline src/eth_rpc/servers/txpool_rpc.rs:10]: `pool_provider` and `PP`?

File: src/providers/alchemy_provider/provider.rs [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+}
+
+impl<P: EthereumProvider> AlchemyDataProvider<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for AlchemyDataProvider<P> {
```
Comment (greged93) [inline src/providers/alchemy_provider/provider.rs:52]: I still don't understand why we need this @eugypalu

File: src/providers/debug_provider/provider.rs [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+}
+
+impl<P: EthereumProvider> DebugDataProvider<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for DebugDataProvider<P> {
```
Comment (greged93) [inline src/providers/debug_provider/provider.rs:68]: same

File: src/providers/pool_provider/provider.rs [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+impl<P: EthereumProvider> PoolDataProvider<P> {
+    pub const fn new(eth_provider: P) -> Self {
+        Self { eth_provider }
+    }
+}
+
+#[async_trait]
+impl<P: EthereumProvider + Send + Sync + 'static> EthereumProvider for PoolDataProvider<P> {
```
Comment (greged93) [inline src/providers/pool_provider/provider.rs:34]: same

**Review by greged93 — COMMENTED** [review]
the organisation of the folders are a little weird, could we have:
- providers:
   - eth_provider/**
   - debug_provider.rs
   - pool_provider.rs
   - alchemy_provider.rs
I don't think we need to have each provider have its own folder for now

File: src/providers/alchemy_provider/provider.rs (hunk 13 above)
Comment (eugypalu) [reply src/providers/alchemy_provider/provider.rs:52]: I think you are right, I will remove it from all three providers

Comment (eugypalu) [reply src/providers/alchemy_provider/provider.rs]: No, sorry, I reviewed the code, and EthereumProvider is used in all three providers:

- in pool_provider to call txpool_content
- in debug to call block_by_hash, etc.
- in alchemy for EthereumErc20

So it seemed like a good solution to add it, but maybe I'm missing something. What do you think?

Comment (greged93) [reply src/providers/alchemy_provider/provider.rs]: afaiu you need an eth_provider but you don't need to implement the provider trait. I will pull the PR and check locally

Comment (greged93) [reply src/providers/alchemy_provider/provider.rs]: I pushed a commit, you can check it, this should work

Comment (eugypalu) [reply src/providers/alchemy_provider/provider.rs]: I saw the commit, now I understand what you meant. Thanks

**Review by greged93 — APPROVED** [review]

**Pull request merged**